### PR TITLE
Issues 060

### PR DESCRIPTION
**Pull request by VolodymyrBabaiev**
Updated scbootstrap tool. Fixed generateCertWithKeyRotationProofInfo information message. Updated error messages. Added duplication check for signer and master keys.

Fixed isNonCeasing missing param error message. Added isNonCeasing to all configuration files.

Updated sc_multiple_pending_certs_non_ceasing.py to use for longer pending distances.

### CODE REVIEW

**Review by i-Alex — APPROVED** [review]

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java [hunk 1]
```diff
@@ -393,27 +402,40 @@ private void processGenerateCertWithKeyRotationProofInfo(JsonNode json) throws E
...
+                return;
+            }
+
+            mastersPublicKeys.add(masterKey);
+            index++;
         }
 
-        assert mastersPublicKeys.size() == signersPublicKeys.size() : "mastersPublicKeys and signersPublicKeys must have the same size";
+        if (mastersPublicKeys.size() != signersPublicKeys.size()) {
+            printGenerateCertWithKeyRotationProofInfoUsageMsg(String.format("an amount of the signer keys and the master keys must be equal."));
```
Comment (rushby) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:418]: This error could be clearer:

"the number of signer keys must be equal to the number of master keys"

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java [hunk 2]
```diff
@@ -393,27 +402,40 @@ private void processGenerateCertWithKeyRotationProofInfo(JsonNode json) throws E
...
-            printGenerateCertWithKeyRotationProofInfoUsageMsg("wrong threshold");
+            printGenerateCertWithKeyRotationProofInfoUsageMsg("threshold is missing or it has unsupported format");
             return;
         }
 
         int threshold = json.get("threshold").asInt();
 
         if (threshold <= 0 || threshold > signersPublicKeys.size()) {
-            printGenerateCertWithKeyRotationProofInfoUsageMsg("wrong threshold: " + threshold);
+            printGenerateCertWithKeyRotationProofInfoUsageMsg("threshold parameter should be greater than 0 and be less or equal to keyCount. Current value: " + threshold);
```
Comment (rushby) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:438]: "less or equal" change to "less than or equal"

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java [hunk 3]
```diff
@@ -393,27 +402,40 @@ private void processGenerateCertWithKeyRotationProofInfo(JsonNode json) throws E
                 return;
             }
 
-            mastersPublicKeys.add(pkNode.asText());
+            String masterKey = pkNode.asText();
+
+            int duplicateIndex = mastersPublicKeys.indexOf(masterKey);
+            if (duplicateIndex != -1) {
+                printGenerateCertWithKeyRotationProofInfoUsageMsg(String.format("duplicated keys in master keys are found. MasterKey with index %d equals to masterKey with index %d", duplicateIndex + 1, index));
```
Comment (rushby) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:409]: Consider changing error to:

"masterKeys contains duplicate values. MasterKey with index %d equals to masterKey with index %d"

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java [hunk 4]
```diff
@@ -374,17 +373,27 @@ private void processGenerateCertWithKeyRotationProofInfo(JsonNode json) throws E
                 return;
             }
 
-            signersPublicKeys.add(pkNode.asText());
+            String pk = pkNode.asText();
+
+            int duplicateIndex = signersPublicKeys.indexOf(pk);
+            if (duplicateIndex != -1) {
+                printGenerateCertWithKeyRotationProofInfoUsageMsg(String.format("duplicated keys in signers keys are found. SignersKey with index %d equals to signersKey with index %d", duplicateIndex + 1, index));
```
Comment (rushby) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:380]: Consider changing error message to

"signersKeys contains duplicate values. SignersKey with index %d is identical to signersKey with index %d"

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java (hunk 3 above)
Comment (rushby) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:409]: Consider changing error to:

"masterKeys contains duplicate values. MasterKey with index %d is identical to masterKey with index %d"

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java (hunk 1 above)
Comment (rushby) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:418]: This error could be clearer:

"the number of signer keys must be equal to the number of master keys"

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java (hunk 2 above)
Comment (rushby) [inline tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java:438]: "less or equal" change to "less than or equal"

File: qa/sc_multiple_pending_certs_non_ceasing.py [hunk 5]
```diff
@@ -110,9 +109,9 @@ class SCMultiplePendingCertsNonCeasing(SidechainTestFramework):
 
     sc_nodes_bootstrap_info = None
     sc_withdrawal_epoch_length = 10  # Note: must be even number
-    total_withdrawal_epochs_number = 5
+    total_withdrawal_epochs_number = 5  # sc_withdrawal_epoch_length must exceed this value at least on 2
```
Comment (rushby) [inline qa/sc_multiple_pending_certs_non_ceasing.py:112]: Should this read: "...at least **_by_** 2"?

File: sdk/src/main/scala/com/horizen/SidechainSettings.scala [hunk 6]
```diff
@@ -25,7 +25,7 @@ case class GenesisDataSettings(scGenesisBlockHex: String,
                                mcNetwork: String,
                                withdrawalEpochLength: Int,
                                initialCumulativeCommTreeHash: String,
-                               isNonCeasing: Boolean = false
+                               isNonCeasing: Boolean
```
Comment (rushby) [inline sdk/src/main/scala/com/horizen/SidechainSettings.scala:28]: Is there a check that this value has been set in the config?

**Review by rushby — COMMENTED** [review]
Reviewed, comments added related to error message content

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java (hunk 4 above)
Comment (VolodymyrBabaiev) [reply tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java]: Done.

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java (hunk 3 above)
Comment (VolodymyrBabaiev) [reply tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java]: Done.

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java (hunk 1 above)
Comment (VolodymyrBabaiev) [reply tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java]: Done.

File: tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java (hunk 2 above)
Comment (VolodymyrBabaiev) [reply tools/sctool/src/main/java/com/horizen/ScBootstrappingToolCommandProcessor.java]: Done.

File: qa/sc_multiple_pending_certs_non_ceasing.py (hunk 5 above)
Comment (VolodymyrBabaiev) [reply qa/sc_multiple_pending_certs_non_ceasing.py]: Done.

File: sdk/src/main/scala/com/horizen/SidechainSettings.scala (hunk 6 above)
Comment (VolodymyrBabaiev) [reply sdk/src/main/scala/com/horizen/SidechainSettings.scala:28]: Yes, there are automatic checks for all settings that don't have default value.

**Review by rushby — APPROVED** [review]

**Pull request merged**